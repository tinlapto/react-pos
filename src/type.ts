export type Item = {
  id: number;
  title: string;
  photo: string;
  price: number;
};

export type Topping = {
  id: number;
  name: string;
};

export type Size = {
  id: number;
  name: string;
};

export type ItemDetail = Item & {
  toppings: Topping[];
  sizes: Size[];
};

export type CartItem = {
  id: number;
  title: string;
  price: number;
  toppings: Topping[];
  sizes: Size[];
  selectedToppings: number[];
  selectedSize: number;
  count: number;
};
