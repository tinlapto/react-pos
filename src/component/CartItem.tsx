import { Button, Col, Popconfirm, Row } from "antd";
import { CloseOutlined } from "@ant-design/icons";
import { CartItem as TypeCartItem } from "../type";

const CartItem = ({
  item,
  onRemove,
}: {
  item: TypeCartItem;
  onRemove: () => void;
}) => {
  const {
    title,
    price,
    count,
    toppings,
    selectedToppings,
    sizes,
    selectedSize,
  } = item;
  return (
    <Row
      style={{
        margin: "5px 0",
        borderBottom: "0.1px solid #eee",
        padding: "5px 0",
      }}
      gutter={10}
    >
      <Col style={{ display: "flex", alignItems: "center" }} span={4}>
        <Popconfirm
          title={"Confirm to remove item?"}
          okText={"Remove"}
          okButtonProps={{ danger: true }}
          onConfirm={onRemove}
        >
          <Button size={"small"} danger icon={<CloseOutlined />} />
        </Popconfirm>
      </Col>
      <Col style={{ textAlign: "right" }} span={16}>
        <div>
          {count} X {title}
        </div>
        <div>
          {toppings
            .filter(({ id }) => selectedToppings.includes(id))
            .map(({ name }, index) => (
              <div key={index} style={{ color: "#aaa", fontSize: 12 }}>
                {name}
              </div>
            ))}
        </div>
        <div>
          <div style={{ color: "#aaa", fontSize: 12 }}>
            {sizes.find(({ id }) => selectedSize === id)?.name}
          </div>
        </div>
      </Col>
      <Col
        style={{
          display: "flex",
          alignItems: "center",
          justifyContent: "right",
        }}
        span={4}
      >
        ${price}
      </Col>
    </Row>
  );
};

export default CartItem;
