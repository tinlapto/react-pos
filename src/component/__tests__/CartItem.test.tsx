import { render, screen, fireEvent } from "@testing-library/react";
import CartItem from "../CartItem";
import { CartItem as TypeCartItem } from "../../type";

const item: TypeCartItem = {
  id: 1,
  title: "test item",
  toppings: [{ id: 1, name: "topping 1" }],
  sizes: [{ id: 1, name: "size 1" }],
  selectedToppings: [1],
  selectedSize: 1,
  count: 1,
  price: 10,
};

describe("CartItem", () => {
  beforeAll(() => {
    // @ts-ignore
    delete window.matchMedia;
    window.matchMedia = (query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    });
  });
  test("render", () => {
    const onRemove = jest.fn();
    const { asFragment } = render(
      <CartItem item={item} onRemove={onRemove} />,
      { container: document.body }
    );
    expect(asFragment()).toMatchSnapshot();

    //  Click X icon
    fireEvent.click(screen.getByRole("img"));

    //  Click confirm remove
    fireEvent.click(screen.getByText("Remove"));
    expect(onRemove).toBeCalled();
  });
});
