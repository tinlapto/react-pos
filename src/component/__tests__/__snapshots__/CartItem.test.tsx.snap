// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem render 1`] = `
<DocumentFragment>
  <div
    class="ant-row"
    style="margin: 5px 0px; border-bottom: 0.1px solid #eee; padding: 5px 0px;"
  >
    <div
      class="ant-col ant-col-4"
      style="padding-left: 5px; padding-right: 5px; display: flex; align-items: center;"
    >
      <button
        class="ant-btn ant-btn-default ant-btn-sm ant-btn-icon-only ant-btn-dangerous"
        type="button"
      >
        <span
          aria-label="close"
          class="anticon anticon-close"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="ant-col ant-col-16"
      style="padding-left: 5px; padding-right: 5px; text-align: right;"
    >
      <div>
        1 X test item
      </div>
      <div>
        <div
          style="color: rgb(170, 170, 170); font-size: 12px;"
        >
          topping 1
        </div>
      </div>
      <div>
        <div
          style="color: rgb(170, 170, 170); font-size: 12px;"
        >
          size 1
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-4"
      style="padding-left: 5px; padding-right: 5px; display: flex; align-items: center; justify-content: right;"
    >
      $10
    </div>
  </div>
</DocumentFragment>
`;
